src/Quantum.jl

```diff
@@ -1,159 +1,94 @@
+inds(tn::AbstractQuantum, ::Val{:at}, site::Site) = inds(Quantum(tn), Val(:at), site)
+tensors(tn::AbstractQuantum, ::Val{:at}, site::Site) = only(tensors(tn; intersects=inds(tn; at=site)))
 
+# `AbstractQuantum` interface
+# TODO would be simpler and easier by overloading `Core.kwcall`? ⚠️ it's an internal implementation detail
 """
-    adjoint(q::Quantum)
+    sites(q::AbstractQuantum)
 
-Returns the adjoint of a [`Quantum`](@ref) Tensor Network; i.e. the conjugate Tensor Network with the inputs and outputs swapped.
+Returns the sites of a [`AbstractQuantum`](@ref) Tensor Network.
 """
-function Base.adjoint(qtn::Quantum)
-    sites = Dict{Site,Symbol}(
-        Iterators.map(qtn.sites) do (site, index)
-            site' => index
-        end,
-    )
-
-    tn = conj(TensorNetwork(qtn))
-
-    # rename inner indices
-    physical_inds = values(sites)
-    virtual_inds = setdiff(inds(tn), physical_inds)
-    replace!(tn, map(virtual_inds) do i
-        i => Symbol(i, "'")
-    end...)
-
-    return Quantum(tn, sites)
+function sites(tn::AbstractQuantum; kwargs...)
+    isempty(kwargs) && return sites(tn, Val(:set), :all)
+    key = only(keys(kwargs))
+    value = values(kwargs)[key]
+    return sites(tn, Val(key), value)
 end
 
-"""
-    ninputs(q::Quantum)
-
-Returns the number of input sites of a [`Quantum`](@ref) Tensor Network.
-"""
-ninputs(q::Quantum) = count(isdual, keys(q.sites))
-
-"""
-    noutputs(q::Quantum)
-
-Returns the number of output sites of a [`Quantum`](@ref) Tensor Network.
-"""
-noutputs(q::Quantum) = count(!isdual, keys(q.sites))
+function nsites(tn::AbstractQuantum; kwargs...)
+    isempty(kwargs) && return nsites(tn, Val(:set), :all)
+    key = only(keys(kwargs))
+    value = values(kwargs)[key]
+    return nsites(tn, Val(key), value)
+end
 
 """
     inputs(q::Quantum)
 
 Returns the input sites of a [`Quantum`](@ref) Tensor Network.
 """
-inputs(q::Quantum) = sort!(collect(filter(isdual, keys(q.sites))))
+# inputs(q::Quantum) = sort!(collect(filter(isdual, keys(q.sites))))
+@deprecate inputs(tn::AbstractQuantum) sites(tn; set=:inputs)
 
 """
     outputs(q::Quantum)
 
 Returns the output sites of a [`Quantum`](@ref) Tensor Network.
 """
-outputs(q::Quantum) = sort!(collect(filter(!isdual, keys(q.sites))))
-
-Base.summary(io::IO, q::Quantum) = print(io, "$(length(q.tn.tensormap))-tensors Quantum")
-Base.show(io::IO, q::Quantum) = print(io, "Quantum (inputs=$(ninputs(q)), outputs=$(noutputs(q)))")
-
-"""
-    sites(q::Quantum)
-
-Returns the sites of a [`Quantum`](@ref) Tensor Network.
-"""
-function sites(tn::Quantum; kwargs...)
-    if isempty(kwargs)
-        collect(keys(tn.sites))
-    elseif keys(kwargs) === (:at,)
-        findfirst(i -> i === kwargs[:at], tn.sites)
-    else
-        throw(MethodError(sites, (Quantum,), kwargs))
-    end
-end
+# outputs(q::Quantum) = sort!(collect(filter(!isdual, keys(q.sites))))
+@deprecate outputs(tn::AbstractQuantum) sites(tn; set=:outputs)
```

Comment:
I don't know why but (same as you yesterday) I don't get any warning about your deprecations here (though it seems to be calling the deprecate.. ) - did you force disable warns somewhere ?


```
julia> inputs(mps1)
Set{Site}()

julia> mps1 isa AbstractQuantum
ERROR: UndefVarError: `AbstractQuantum` not defined in `Main`
Suggestion: check for spelling errors or missing imports.
Stacktrace:
 [1] top-level scope
   @ REPL[13]:1

julia> mps1 isa Tenet.AbstractQuantum
true

julia> outputs(mps1)
Set{Site} with 5 elements:
  3
  5
  1
  2
  4

julia> ninputs(mps1)
0

julia> @which ninputs(mps1)
ninputs(tn::Tenet.AbstractQuantum)
     @ Tenet deprecated.jl:103
```

Reply:
Yeah, it's weird. This also happens to me but shouldn't be like this.

Reply:
ok, apparently you need to start julia with `julia --depwarn=yes` or [they don't show](https://invenia.github.io/blog/2022/06/17/deprecating-in-julia/#:~:text=Deprecation%20warnings%20are%20suppressed%20in,starting%20a%20new%20Julia%20session.)

meh 

```
julia> mps1 = rand(Chain, Open, State; n=5, d=2, χ=5 )
MPS (inputs=0, outputs=5)

julia> inputs(mps1)
┌ Warning: `inputs(tn::AbstractQuantum)` is deprecated, use `sites(tn; set = :inputs)` instead.
│   caller = top-level scope at REPL[5]:1
└ @ Core REPL[5]:1
Set{Site}()
```

Reply:
ahh that makes sense. I think that the Julia subprocesses spawned by `test` do have that flag activated.

Reply:
Thanks! @starsfordummies